File: DP/Longest Palindrome Substring/LPS.cpp
#include <iostream>
#include <cstring>

using namespace std;

const int MAX = 1010;

int memo[MAX][MAX]; 
int lps(const string& s, int i, int j)
{

    if (i > j)
        return 0;

    if (memo[i][j] > -1)
        return memo[i][j];

    if (s[i] == s[j])
    {
        int equalCharacters = 2 - (i == j);
        return memo[i][j] = equalCharacters + lps(s, i + 1, j - 1);
    }
    return memo[i][j] = max( lps(s, i + 1, j), lps(s, i, j - 1) );
}

int longest_palindrome(const string& s)
{

    memset(memo, -1, sizeof memo);
    return lps(s, 0, s.length() - 1);
}

int main()
{

    cout << longest_palindrome("bbabcbcab") << '\n';
    cout << longest_palindrome("abbaab") << '\n';
    cout << longest_palindrome("opengenus") << '\n'; 

    return 0;
}

File: DP/Longest Palindrome Substring/LPS.py
def longest_palindrome(text):
    
    length = len(text)
    palindromes_lengths = [1] * length
    for fgap in range(1, length):
        pre = palindromes_lengths[fgap]
        for rgap in reversed(range(0, fgap)):
            tmp = palindromes_lengths[rgap]
            if text[fgap] == text[rgap]:
                if rgap + 1 > fgap - 1:
                    palindromes_lengths[rgap] = 2
                else:
                    palindromes_lengths[rgap] = 2 + pre
            else:
                palindromes_lengths[rgap] = max(palindromes_lengths[rgap + 1],
                                                palindromes_lengths[rgap])
            pre = tmp
    return palindromes_lengths[0]

def main():
    print(longest_palindrome('bbabcbcab')) 
    print(longest_palindrome('abbaab'))    
    print(longest_palindrome('opengenus')) 

if __name__ == '__main__':
    main()
